MM.lua

```diff
@@ -0,0 +1,88 @@
+--[[ Module to perform matrix multiplication on two minibatch inputs,
+     producing a minibatch.
+]]
+
+local MM, parent = torch.class('nn.MM', 'nn.Module')
+
+--[[ The constructor takes two optional arguments, specifying whether or not transpose
+     any of the input matrices before perfoming the multiplication.
+]]
+function MM:__init(transA, transB)
+  parent.__init(self)
+
+  self.transA = transA or false
+  self.transB = transB or false
+
+  self.gradInput = {torch.Tensor(), torch.Tensor()}
+end
+
+function MM:updateOutput(input)
+  assert(#input == 2, 'input must be a pair of minibatch matrices')
+  local a, b = unpack(input)
+  assert(a:nDimension() == 2 or a:nDimension() == 3, 'arguments must be a 2D or 3D Tensor')
+
+  if a:nDimension() == 2 then
+    assert(b:nDimension() == 2, 'arg 2 must be a 2D Tensor')
+
+    if self.transA then a = a:t() end
+    if self.transB then b = b:t() end
+    assert(a:size(2) == b:size(1), 'matrix sizes do not match')
+
+    self.output:resize(a:size(1), b:size(2))
+    self.output:mm(a, b)
+  else
+    assert(b:nDimension() == 3, 'arg 2 must be a 3D Tensor')
+    assert(a:size(1), b:size(2), 'inputs must contain the same number of minibatches')
+
+    if self.transA then a = a:transpose(2, 3) end
+    if self.transB then b = b:transpose(2, 3) end
+    assert(a:size(3) == b:size(2), 'matrix sizes do not match')
+
+    self.output:bmm(a, b)
+  end
+
+  return self.output
+end
+
+function MM:updateGradInput(input, gradOutput)
+  assert(#input == 2, 'input must be a pair of minibatch matrices')
+  local a, b = unpack(input)
+  self.gradInput[1]:resizeAs(a)
+  self.gradInput[2]:resizeAs(b)
+
+  assert(gradOutput:nDimension() == 2 or gradOutput:nDimension() == 3, 'arguments must be a 2D or 3D Tensor')
+
+  local h_dim, w_dim, f
+  if gradOutput:nDimension() == 2 then
+    assert(a:nDimension() == 2, 'arg 2 must be a 2D Tensor')
```

Comment:
Could you make the messages here more specific? Like arg 2a vs arg 2b or something.